neo/Helper.cs

```diff
@@ -320,5 +320,18 @@ public static IConfigurationRoot LoadConfig(string config)
                 .AddJsonFile(configFile, true)
                 .Build();
         }
+
+        public static T StrictEnum<T>(byte value) where T : Enum
+        {
+            T state = (T)Enum.ToObject(typeof(T), value);
+            if (!EnumValueCache<T>.definedValues.Contains(state))
+                throw new FormatException();
+            return state;
+        }
+
+        private static class EnumValueCache<T> where T : Enum
+        {
+            public static readonly HashSet<T> definedValues = new HashSet<T>((T[])Enum.GetValues(typeof(T)));
+        }
```

Comment:
I prefer to use `Enum.IsDefined()`.

Reply:
Can you explain why? It is 2x slower than with this hashset. See https://github.com/neo-project/neo/issues/1231#issuecomment-552860288

Reply:
My test:

```c#
enum TestEnum : byte
{
    Zero,
    One,
    Two,
    Three,
    Four,
    Five,
    Six,
    Seven,
    Eight,
    Nine,
}

static class EnumValueCache<T> where T : Enum
{
    public static readonly HashSet<T> DefinedValues = new HashSet<T>((T[])Enum.GetValues(typeof(T)));
}

static T StrictEnum<T>(byte value) where T : Enum
{
    T state = (T)Enum.ToObject(typeof(T), value);
    if (!EnumValueCache<T>.DefinedValues.Contains(state))
        throw new FormatException();
    return state;
}

static void Main()
{
    Stopwatch stopwatch = new Stopwatch();
    TestEnum five = TestEnum.Five;
    byte fiveValue = (byte)TestEnum.Five;

    stopwatch.Start();
    for (int i = 0; i < 100000; i++)
    {
        Enum.IsDefined(typeof(TestEnum), five);
    }
    stopwatch.Stop();
    Console.WriteLine($"IsDefined: {stopwatch.Elapsed.TotalMilliseconds} ms");

    stopwatch.Restart();
    for (int i = 0; i < 100000; i++)
    {
        Enum.IsDefined(typeof(TestEnum), TestEnum.Five);
        StrictEnum<TestEnum>(fiveValue);
    }
    stopwatch.Stop();
    Console.WriteLine($"StrictEnum: {stopwatch.Elapsed.TotalMilliseconds} ms");
}
```

IsDefined: 41.8113 ms
StrictEnum: 121.79360000000001 ms

Reply:
I don't think this is a completely fair comparison. We're comparing a generic function against an inlined implementation that does not have the function call overhead,  that only deals with 1 enum type and doesn't check if it needs to raise any exceptions (and maybe even benefits from internal optimizations because it's all in one loop).

Here are my results when I make the code around `Enum.IsDefined` generic and put it into its own helper function. Reducing boilerplate code was part of the idea of the PR.

**with fixed value**
`Enum.IsDefined` best case - 26.602ms (while actually averaging around 28.xx)
`StrictEnum` worst case - 23.6009ms
![](https://i.imgur.com/EkgLsyz.png)

**with changing value**
`Enum.IsDefined` best case - 32.0974ms (while actually averaging around 33.xx)
`StrictEnum` worst case - 23.966ms
![](https://i.imgur.com/61Zn1UI.png)

It is not the 2x difference that my previous test showed, but it is still faster. The difference in the old results might be caused by the machine used (old laptop then vs desktop now).

If we still prefer `Enum.IsDefined` then no problem but to me it seems like the hashset solution is the faster and better option because neo is a long running program with varying inputs to be casted.